packages/mui-base/src/NumberInput/NumberInput.test.tsx

```diff
@@ -0,0 +1,469 @@
+        />,
+      );
+
+      const input = getByTestId('input') as HTMLInputElement;
+
+      await user.click(input);
+
+      await user.keyboard('[Home]');
+      expect(handleValueChange.args[0][1]).to.equal(50);
+      expect(input.value).to.equal('50');
+    });
+
+    it('sets value to min when End is pressed', async () => {
+      const user = userEvent.setup();
+      const handleValueChange = spy();
+
+      const { getByTestId } = render(
+        <NumberInput
+          defaultValue={10}
+          min={1}
+          onValueChange={handleValueChange}
+          slotProps={{ input: { 'data-testid': 'input' } } as any}
+        />,
+      );
+
+      const input = getByTestId('input') as HTMLInputElement;
+
+      await user.click(input);
+
+      await user.keyboard('[End]');
+      expect(handleValueChange.args[0][1]).to.equal(1);
+      expect(input.value).to.equal('1');
+    });
+
+    it('sets value to min when the input has no value and ArrowUp is pressed', async () => {
+      const user = userEvent.setup();
+      const handleValueChange = spy();
+
+      const { getByTestId } = render(
+        <NumberInput
+          min={5}
+          onValueChange={handleValueChange}
+          slotProps={{ input: { 'data-testid': 'input' } } as any}
+        />,
+      );
+
+      const input = getByTestId('input') as HTMLInputElement;
+
+      await user.click(input);
+
+      await user.keyboard('[ArrowUp]');
+      expect(handleValueChange.args[0][1]).to.equal(5);
+      expect(input.value).to.equal('5');
+    });
+
+    it('sets value to max when the input has no value and ArrowDown is pressed', async () => {
+      const user = userEvent.setup();
+      const handleValueChange = spy();
+
+      const { getByTestId } = render(
+        <NumberInput
+          max={9}
+          onValueChange={handleValueChange}
+          slotProps={{ input: { 'data-testid': 'input' } } as any}
+        />,
+      );
+
+      const input = getByTestId('input') as HTMLInputElement;
+
+      await user.click(input);
+
+      await user.keyboard('[ArrowDown]');
+      expect(handleValueChange.args[0][1]).to.equal(9);
+      expect(input.value).to.equal('9');
+    });
+  });
+
+  describe('prop: readOnly', () => {
+    it('stepper buttons should not be in the tab order when readOnly is false', async () => {
+      const user = userEvent.setup();
+
+      const { getByTestId } = render(
+        <NumberInput slotProps={{ input: { 'data-testid': 'input' } } as any} />,
+      );
+
+      const input = getByTestId('input') as HTMLInputElement;
+      expect(document.activeElement).to.equal(document.body);
+
+      await user.keyboard('[Tab]');
+      expect(document.activeElement).to.equal(input);
+
+      await user.keyboard('[Tab]');
+      expect(document.activeElement).to.equal(document.body);
+    });
+
+    it('tab order should be increment button, decrement button when readOnly is true', async () => {
```

Comment:
What's the rationale behind this? If a read-only input is not focusable, there's no way for the screen reader to announce its value.

Reply:
> The text field is usually the only focusable component because the increase and decrease functions are keyboard accessible via arrow keys

It's kind of the opposite of this from [here](https://www.w3.org/WAI/ARIA/apg/patterns/spinbutton/), it's not uncommon to see "number inputs" where only the stepper buttons are interact-able and the number is just a display:

![Screenshot 2023-05-22 at 5 45 26 PM](https://github.com/mui/material-ui/assets/4997971/f53032dc-f52c-4538-8438-0b9204986acd)

> there's no way for the screen reader to announce its value

Yes! I think this is the most common shortcoming of these "stepper only" number inputs but I'm not 100% sure what the best thing to do is yet.

But perhaps hanging this all on the `readOnly` prop is not the right start, as it may confuse users with the `readonly` DOM attribute which only really concerns the `<input>`, what do you think? @michaldudak

Reply:
Yeah, my initial thought was that something did not work - even though I set `readOnly`, the component was still editable. Perhaps something along the lines of `disableTextInput` would be more intuitive?
As for a11y, it could make sense to implement both buttons in a single focus group (as in https://www.w3.org/WAI/ARIA/apg/patterns/spinbutton/examples/datepicker-spinbuttons/). This way, we could add ARIA attributes to indicate the current value. But this would be a pretty big change when compared to the current state, right? We could discuss and tackle it in a separate PR.